internal/database/pebble.go

```diff
@@ -0,0 +1,146 @@
+// Copyright 2023 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package database
+
+import (
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/ChainSafe/gossamer/internal/log"
+	"github.com/cockroachdb/pebble"
+	"github.com/cockroachdb/pebble/vfs"
+)
+
+var logger = log.NewFromGlobal(log.AddContext("internal", "database"))
+var _ Database = (*pebbleDB)(nil)
+
+var ErrNotFound = pebble.ErrNotFound
+
+type pebbleDB struct {
+	path string
+	db   *pebble.DB
+}
+
+func NewPebble(path string, inMemory bool) (*pebbleDB, error) {
+	opts := &pebble.Options{}
+	if inMemory {
+		opts = &pebble.Options{FS: vfs.NewMem()}
+	} else {
+		if err := os.MkdirAll(path, os.ModePerm); err != nil {
+			return nil, err
+		}
+	}
+
+	db, err := pebble.Open(path, opts)
+	if err != nil {
+		return nil, fmt.Errorf("oppening pebble db: %w", err)
+	}
+
+	return &pebbleDB{path, db}, nil
+}
+
+func (p *pebbleDB) Path() string {
+	return p.path
+}
+
+func (p *pebbleDB) Put(key, value []byte) error {
+	err := p.db.Set(key, value, &pebble.WriteOptions{})
+	if err != nil {
+		return fmt.Errorf("writing 0x%x with value 0x%x to database: %w",
+			key, value, err)
+	}
+	return nil
+}
+
+func (p *pebbleDB) Get(key []byte) (value []byte, err error) {
+	value, closer, err := p.db.Get(key)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := closer.Close(); err != nil {
+		return nil, fmt.Errorf("closing after get: %w", err)
+	}
+
+	valueCpy := make([]byte, len(value))
+	copy(valueCpy[:], value[:])
+	return valueCpy, err
+}
+
+func (p *pebbleDB) Has(key []byte) (exists bool, err error) {
+	value, closer, err := p.db.Get(key)
+	if err != nil {
+		if errors.Is(err, pebble.ErrNotFound) {
+			return false, nil
+		}
+
+		return false, err
+	}
+
+	if err := closer.Close(); err != nil {
+		return false, fmt.Errorf("closing after get: %w", err)
+	}
+
+	return value != nil, err
+}
+
+func (p *pebbleDB) Del(key []byte) error {
+	err := p.db.Delete(key, &pebble.WriteOptions{})
+	if err != nil {
+		return err
```

Comment:
wrap error?

Reply:
I don't think wrapping the error here will be useful since this method is just a proxy to the `db.Delete`, I mean, what kind of context we can add to the error message that is useful to the caller?